zigup.zig

```diff
@@ -421,19 +425,75 @@ fn listCompilers(allocator: *Allocator) !void {
     }
 }
 
-fn printDefaultCompiler(allocator: *Allocator) !void {
+fn cleanCompilers(allocator: *Allocator) !void {
+    const install_dir_string = try getInstallDir(allocator, .{ .create = true });
+    defer allocator.free(install_dir_string);
+    // getting the current compiler
+    const default_comp_opt = try getDefaultCompiler(allocator);
+    defer if (default_comp_opt) |_| allocator.free(default_comp_opt.?);
+
+    var install_dir = std.fs.cwd().openDir(install_dir_string, .{ .iterate = true }) catch |e| switch (e) {
+        error.FileNotFound => return,
+        else => return e,
+    };
+    defer install_dir.close();
+    var it = install_dir.iterate();
+    while (try it.next()) |entry| {
+        if (entry.kind != .Directory)
+            continue;
+        if (std.mem.endsWith(u8, entry.name, ".installing"))
+            continue;
+
+        if (default_comp_opt) |default_comp| {
+            if (mem.eql(u8, default_comp, entry.name)) {
+                std.debug.warn("Not deleting the default compiler: {}\n", .{default_comp});
+                continue;
+            }
+        }
+        // if its in the compilers to keep, skip it
+
+        {
+            const look_for_keep_path = try std.fs.path.join(allocator, &[_][]const u8{ install_dir_string, entry.name, "keep" });
+            defer allocator.free(look_for_keep_path);
+            if (std.fs.openFileAbsolute(look_for_keep_path, .{ .read = false })) |keep_file| {
+                keep_file.close();
+                std.debug.warn("Not cleaning '{}' because of keep file.\n", .{entry.name});
+                continue;
+            } else |e| switch (e) {
+                error.FileNotFound => {},
+                else => return e,
+            }
+        }
+        // if we didn't continue at all then actually delete the compiler
+        const abs_path_to_delete = try std.fs.path.join(allocator, &[_][]const u8{ install_dir_string, entry.name });
+        defer allocator.free(abs_path_to_delete);
+        try loggyDeleteTreeAbsolute(abs_path_to_delete);
+    }
+}
+
+fn getDefaultCompiler(allocator: *Allocator) !?[]const u8 {
     const pathLink = try makeZigPathLinkString(allocator);
     defer allocator.free(pathLink);
-    var targetPathBuffer: [std.fs.MAX_PATH_BYTES]u8 = undefined;
-    if (std.fs.readLinkAbsolute(pathLink, &targetPathBuffer)) |targetPath| {
-        std.debug.warn("{}", .{std.fs.path.basename(std.fs.path.dirname(std.fs.path.dirname(targetPath).?).?)});
+    // var targetPathBuffer: [std.fs.MAX_PATH_BYTES]u8 = undefined;
```

Comment:
Why change this?

Reply:
It returns a stack allocated slice, so I had to manually allocate it. I am going to change this to inline fn so that it can do this.

Reply:
I changed it to take a buffer after chatting with people on irc about best way to do this.

Reply:
gotcha I see now